src/symbols.c

```diff
@@ -1404,90 +1404,28 @@ static guint get_tag_class(const TMTag *tag)
-static void goto_popup_position_func(GtkMenu *menu, gint *x, gint *y, gboolean *push_in, gpointer data)
+/* opens menu at caret position */
+static void show_menu_at_caret(GtkMenu* menu, ScintillaObject *sci)
 {
-	gint line_height;
-	GdkScreen *screen = gtk_widget_get_screen(GTK_WIDGET(menu));
-	gint monitor_num;
-	GdkRectangle monitor;
-	GtkRequisition req;
-	GdkEventButton *event_button = g_object_get_data(G_OBJECT(menu), "geany-button-event");
-
-	if (event_button)
-	{
-		/* if we got a mouse click, popup at that position */
-		*x = (gint) event_button->x_root;
-		*y = (gint) event_button->y_root;
-		line_height = 0; /* we don't want to offset below the line or anything */
-	}
-	else /* keyboard positioning */
-	{
-		ScintillaObject *sci = data;
-		GdkWindow *window = gtk_widget_get_window(GTK_WIDGET(sci));
-		gint pos = sci_get_current_position(sci);
-		gint line = sci_get_line_from_position(sci, pos);
-		gint pos_x = SSM(sci, SCI_POINTXFROMPOSITION, 0, pos);
-		gint pos_y = SSM(sci, SCI_POINTYFROMPOSITION, 0, pos);
-
-		line_height = SSM(sci, SCI_TEXTHEIGHT, line, 0);
-
-		gdk_window_get_origin(window, x, y);
-		*x += pos_x;
-		*y += pos_y;
-	}
-
-	monitor_num = gdk_screen_get_monitor_at_point(screen, *x, *y);
-
-	gtk_widget_get_preferred_size(GTK_WIDGET(menu), NULL, &req);
-
-#if GTK_CHECK_VERSION(3, 4, 0)
-	gdk_screen_get_monitor_workarea(screen, monitor_num, &monitor);
-#else
-	gdk_screen_get_monitor_geometry(screen, monitor_num, &monitor);
-#endif
-
-	/* put on one size of the X position, but within the monitor */
-	if (gtk_widget_get_direction(GTK_WIDGET(menu)) == GTK_TEXT_DIR_RTL)
-	{
-		if (*x - req.width - 1 >= monitor.x)
-			*x -= req.width + 1;
-		else if (*x + req.width > monitor.x + monitor.width)
-			*x = monitor.x;
-		else
-			*x += 1;
-	}
-	else
-	{
-		if (*x + req.width + 1 <= monitor.x + monitor.width)
-			*x = MAX(monitor.x, *x + 1);
-		else if (*x - req.width - 1 >= monitor.x)
-			*x -= req.width + 1;
-		else
-			*x = monitor.x + MAX(0, monitor.width - req.width);
-	}
-
-	/* try to put, in order:
-	 * 1. below the Y position, under the line
-	 * 2. above the Y position
-	 * 3. within the monitor */
-	if (*y + line_height + req.height <= monitor.y + monitor.height)
-		*y = MAX(monitor.y, *y + line_height);
-	else if (*y - req.height >= monitor.y)
-		*y = *y - req.height;
-	else
-		*y = monitor.y + MAX(0, monitor.height - req.height);
-
-	*push_in = FALSE;
+	GdkWindow *window = gtk_widget_get_window(GTK_WIDGET(sci));
+	gint pos = sci_get_current_position(sci);
+	gint line = sci_get_line_from_position(sci, pos);
+	gint line_height = SSM(sci, SCI_TEXTHEIGHT, line, 0);
+	gint x = SSM(sci, SCI_POINTXFROMPOSITION, 0, pos);
+	gint y = SSM(sci, SCI_POINTYFROMPOSITION, 0, pos);
+	gint pos_next = sci_get_position_after(sci, pos);
+	gint char_width = 0;
+	if (sci_get_line_from_position(sci, pos_next) == line)
+		char_width = SSM(sci, SCI_POINTXFROMPOSITION, 0, pos_next) - x;
```

Comment:
As I'm a sucker for weird corner cases, this actually does not work if line wrapping is enabled and the caret is right before the wrap position (popup would be anchored to the start of the next line).

Something like this works though:
```suggestion
	/* if it's on the same Y (same line and not after wrapping), diff the X */
	if (SSM(sci, SCI_POINTYFROMPOSITION, 0, pos_next) == y)
		char_width = SSM(sci, SCI_POINTXFROMPOSITION, 0, pos_next) - x;
	/* otherwise, if it's not at the end of the document, use the character's width
	 * -- which would work in case above, but it feels slower (I didn't get numbers) */
	else if (pos_next > pos)
	{
		gint style = SSM(sci, SCI_GETSTYLEAT, 0, pos);
		gchar *text = sci_get_contents_range(sci, pos, pos_next);
		char_width = sci_text_width(sci, style, text);
		g_free(text);
	}
```

Reply:
Maybe it was mentioned somewhere in the previous conversation but what's the point of calculating `char_width`? Couldn't the rect width just be always `1`? Clearly the height is important so the menu is below the line but with width `1` the menu pops up from the gap between characters which looks fine and not unexpected IMO.

Reply:
> Another possible improvement would be doing something similar in the X direction, although there's no actual issue with it. However, as is, if there isn't enough room on the right, it would then popup on the left of the caret, not under the character at the caret. It might be a nitpick, but it looks a tiny bit off with caret in overwrite mode. Yet, I don't think the previous code supported this.

OK, found it.

Reply:
But really, I'd be for the "lazy" solution here, it's not worth complicating the code too much I think.

Reply:
> But really, I'd be for the "lazy" solution here, it's not worth complicating the code too much I think.

Which one do you refer to as "lazy"?
* current (misbehaving with edge case of wrapping)
* `SCI_POINTYFROMPOSITION`-based one (not perfect for the mentioned edge case, but not bad)
* `SCI_TEXTWIDTH`-base one (probably good for all cases,but possibly slower)

I agree that all the shenanigans might not be needed here, and possibly either solution is OK.  I have a preference for one that doesn't bluntly misbehaves in the edge case, but it's an *very* edge case though…

Reply:
> Which one do you refer to as "lazy"?

Using `char_width = 0` and not calculating it.

Reply:
Ah… I feel it looks bad if cursor is in block mode (or overwrite) and the popup is placed on the left (when there isn't enough room on the right).  IMO, *one* of the solutions (preferably not the first) is better.  But I won't die on that hill :slightly_smiling_face:

Reply:
I don't see what the problem is in this case - this is what happens when I invoke it on `i` and move the window to the right so the popup appears on the left:

<img width="482" alt="Screenshot 2023-10-10 at 12 30 03" src="https://github.com/geany/geany/assets/713965/c99ec979-a4a9-4ca4-b514-2c60d5aff7f5">

The underline cursor of the insert mode disappears when the popup shows.

Reply:
This is with the current state of this PR right, not with just `char_width=0`, is it?

The problem with the current version is only if you have wrapping enabled and it wraps after the next character (e.g. if wrapping wrapped right after the `i` (not even having the space after). What you'd see is the popup on the far left of the editor.

Reply:
> This is with the current state of this PR right, not with just char_width=0, is it?

Yes it is (to be precise, with `char_width=1`, I wasn't sure if 0-width bounding box works alright in this case).

> The problem with the current version is only if you have wrapping enabled and it wraps after the next character (e.g. if wrapping wrapped right after the i (not even having the space after). What you'd see is the popup on the far left of the editor.

I understand the current version problem but I still don't see the problem with `char_width=0`.

Reply:
> Yes it is

To be clear, I meant the screenshot is with the `char_width=1` version.

Reply:
With `char_width=1` (and none of the extra tests), I get this:
![offset-gotopopup](https://github.com/geany/geany/assets/793526/3ca422c3-9203-476f-9c69-a7ee74c11ac0)
Caret is on the last `e` (if in block mode, it is over it, in beam mode right before).

Reply:
Yeah, and what's the problem here? You would like to have the popup's right corner one character to the right? And how is it related to the shape of the caret?

Basically Scintilla internally always has caret between letters and only draws it over the next letter when for instance in a block mode (something I was fighting with in my vimode plugin because vi itself has caret on letters and sometimes this leads to some surprises in character number calculations). So caret over "e" in block mode is actually caret before "e".

But I still don't see the "wrongness" here - the popup jumps from the bottom-left corner of the block mode caret both when there is enough space on the right and also when there isn't and the popup is on the left, so nothing unexpected I'd say. Also this is a rare case where two conditions have to be met - not enough space and a block caret (and I never use overwrite mode, I just had to learn how to invoke "insert" on the mac :-).

But don't hesitate to implement it your way if this is something that bothers you :-).

Reply:
I guess I feel like the popup should be under the letter the caret is at (not sure what are the real differences between having the caret over or before a character though, offsets should be pretty much the same… but this is kind of irrelevant to the feel :))

> But don't hesitate to implement it your way if this is something that bothers you :-).

You nailed it I guess :D

I went ahead and implemented the more basic version that is similar to what was there but don't produce incorrect results on the wrapping corner case.  So I'm happier, and you're not too unhappy because it's not as complicated as it could be :grin:

Reply:
> I went ahead and implemented the more basic version that is similar to what was there but don't produce incorrect results on the wrapping corner case. So I'm happier, and you're not too unhappy because it's not as complicated as it could be 

The best of both worlds indeed, LGTM :-)